Media: add a 'Reverse Order' button to the gallery editor dialog

alisterscott opened the pull request:
This PR seeks to add a ‘Reverse Order’ button to the add new gallery dialog (similar to wp-admin)

Fixes #339 

**Outstanding issues:**

1. Needs design review - I think it makes sense to have the button with the other inputs on the right hand side instead of just in the Edit component
2. `warning.js:36 Warning: Failed prop type: Required prop `legend` was not specified in `EditorMediaModalFieldset`` I'm not sure we need a legend for this button, and I don't know how to supress this warning
3. No unit tests for the new functionality - since no unit tests exist for this existing component I'm not sure where to start on this

cc: @designsimply for testing 😊

**Before**

No 'Reverse Order' option

<img width="817" alt="before calypso" src="https://cloud.githubusercontent.com/assets/128826/19714311/7e27b9be-9b90-11e6-9319-57be0a23c3c2.png">

**After**

![reverse images](https://cloud.githubusercontent.com/assets/128826/19750444/722a321e-9c34-11e6-838f-312d8f4dcfb5.gif)


**WP-Admin**

<img width="782" alt="wp-admin reorder" src="https://cloud.githubusercontent.com/assets/128826/19714248/f25a1684-9b8f-11e6-8fba-5775c4c316cd.png">

alisterscott commented:
Thanks to @jordwest for his feedback/input so far

folletto commented:
> Needs design review - I think it makes sense to have the button with the other inputs on the right hand side instead of just in the Edit component

Agreed, but I think the button should have a different structure / placement.

The priority for that feature is rather minor compared to other controls, so we should move that down. 

The best idea is probably to turn the "Random order" check, which is already a bit odd without label, to a dropdown:

<img width="256" alt="screen shot 2016-10-26 at 11 06 30" src="https://cloud.githubusercontent.com/assets/4389/19719922/490ec7d2-9b6c-11e6-899a-c1194e989c41.png">

lancewillett commented:
Side comment: I love the numbered photos for testing ordering. 💯

lancewillett commented:
+1 to Davide's idea of combining with other ordering controls.

aduth commented on client/post-editor/media-modal/gallery/fields.jsx:
```diff
@@ -15,15 +16,17 @@ import EditorMediaModalFieldset from '../fieldset';
 
-export default React.createClass( {
+export const EditorMediaModalGalleryFields = React.createClass( {
 	displayName: 'EditorMediaModalGalleryFields',
```
When assigning `React.createClass` to a variable, you can remove `displayName`, as it's inferred automatically via [`babel-plugin-react-display-name`](https://www.npmjs.com/package/babel-plugin-react-display-name).

alisterscott replied:
Thanks, I was wondering if we needed that anymore. Removed now.

aduth commented on client/post-editor/media-modal/gallery/fields.jsx:
```diff
@@ -95,6 +98,10 @@ export default React.createClass( {
 	},
 
+	updateReverseOrder() {
+		this.props.onReverse();
```
I'd suggest either:

- Checking that the `onReverse` prop exists before calling on it
- Providing a `noop` (either `() => {}` or [Lodash's `_.noop`](https://lodash.com/docs/4.16.4#noop) as a default prop
- Marking the prop as `.isRequired` in `propTypes`

My preference is toward the second

alisterscott replied:
> Providing a noop (either () => {} or Lodash's _.noop as a default prop

Thanks, I did this.

aduth commented on client/post-editor/media-modal/gallery/fields.jsx:
```diff
@@ -144,13 +151,26 @@ export default React.createClass( {
+
+		return (
+			<EditorMediaModalFieldset>
+				<Button onClick={ this.updateReverseOrder } disabled={ this.props.settings.orderBy === 'rand' } >
```
Do you need `updateReverseOrder` as a function on the instance, or could you just pass `this.props.onReverse` directly? This would also help avoid the need for any checking of truthiness on the prop, as `<Button />` is responsible for handling this.

alisterscott replied:
Thanks, that makes sense: it's not needed.

aduth commented on client/post-editor/media-modal/gallery/fields.jsx:
```diff
@@ -144,13 +151,26 @@ export default React.createClass( {
 
 		return (
 			<div className="editor-media-modal-gallery__fields">
+				{ this.renderReverseOrderButton() }
```
I've started encouraging more rendering inline in the `render` function when possible, as it keeps all render logic colocated. Here, since `renderReverseOrderButton` is simply returning another element, it would be straightforward to move that element in place of the function call here directly. Only in cases where there's conditional render logic does it become more complicated.

alisterscott replied:
Thanks, moved this content into render

aduth commented on client/post-editor/media-modal/gallery/fields.jsx:
```diff
@@ -144,13 +151,26 @@ export default React.createClass( {
 		return (
 			<div className="editor-media-modal-gallery__fields">
+				{ this.renderReverseOrderButton() }
 				{ this.renderDropdown( this.translate( 'Layout' ), types, 'type' ) }
```
I might recommend that you consider updating `this.translate` to `this.props.translate` in the same pull request if you're introducing `localize`.

alisterscott replied:
I thought about that but I didn't want to make this PR too big. Is it ok if I do a separate PR immediately following this to refactor these?

aduth commented on client/post-editor/media-modal/gallery/index.jsx:
```diff
@@ -114,6 +114,13 @@ const EditorMediaModalGallery = React.createClass( {
 	},
 
+	reverseOrder( ) {
+		const reverseItems = reverse( this.props.settings.items.slice( 0 ) );
```
Nice catch on the [mutation note on `_.reverse`](https://lodash.com/docs/4.16.4#reverse) and the need to create a clone of the array 👍 

You could alternatively use newer ES6 semantics to make it more condensed:

```js
const reverseItems = reverse( [ ...this.props.settings.items ] );
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator

alisterscott replied:
Thanks for the tip - updated 😊

aduth commented on client/post-editor/media-modal/gallery/index.jsx:
```diff
@@ -136,7 +143,8 @@ const EditorMediaModalGallery = React.createClass( {
 							onUpdateSetting={ this.updateSetting }
-							numberOfItems={ items.length } />
+							numberOfItems={ items.length }
+							onReverse = { this.reverseOrder } />
```
Extra spaces around the `=` should be removed.

alisterscott replied:
Thanks, removed

alisterscott commented:
Thanks for your feedback @folletto!

> The priority for that feature is rather minor compared to other controls, so we should move that down.

Sure, moved to the bottom now:

<img width="238" alt="screen shot 2016-10-27 at 9 54 19 am" src="https://cloud.githubusercontent.com/assets/128826/19749302/6cf79434-9c2b-11e6-820e-e6d530c46c6f.png">

> The best idea is probably to turn the "Random order" check, which is already a bit odd without label, to a dropdown

I had a chat to @lancewillett about this and it raises a whole lot of additional questions, like what is 'normal' order - is this the 'initial' order or the uploaded order, and what happens when you manually reorder images then choose normal or reverse. To keep this PR small I'd raised this idea as a separate issue which we can work through some ideas/designs for this to implement separately to this. See: #8967

alisterscott commented:
@aduth: thanks for your review. I've addressed all the issues (besides this this.props.translate which I was planning to do as a separate PR straight after this one) - so would you mind re-reviewing this? Thanks

folletto commented:
> I had a chat to @lancewillett about this and it raises a whole lot of additional questions, like what is 'normal' order - is this the 'initial' order or the uploaded order, and what happens when you manually reorder images then choose normal or reverse. To keep this PR small I'd raised this idea as a separate issue which we can work through some ideas/designs for this to implement separately to this.

I don't think that's an issue, because the same question can be raised with the button. In other words: if the issue isn't there with the button, it shouldn't be there with the dropdown. If it's a reason to not make the dropdown, then it's a reason to not do the button either. That sidebar is becoming very crowded.

Ok, I'll move the discussion on the other thread.

alisterscott commented:
> but the "let's not merge this" is even more relevant after that.

Closing.

lancewillett commented:
@folletto What amount of weight do you assign this being a parity with wp-admin? That's part of the reason we split out the small change from the bigger discussion and further improvement? It's <a href="https://github.com/Automattic/wp-calypso/milestone/129">a big list (19 open today)</a> and this change would reduce that disconnect a bit, on its own.

I think we should push the change with just the button for parity, then explore the better UX as a second PR.

folletto commented:
It's not about weight parity, is about adding the button in the wrong part of the UI, a UI that is already full of elements. Adding features and _then_ doing a design for them is the wrong process. Any element on that list should have a design or a design review before it gets developed.

On this specific issue, my design suggestion is in #8967 and the following comment adds on the approach we should take. As you can see there, "feature parity" might not be the full answer to the real issue.
